### ISSUE TITLE
Different behaviour in javascript toolkits with and without optimizations

### ISSUE THREAD

**Issue by JRegimbal** [rism-ch/verovio#897]
A bug appears in the javascript toolkit when built with the default `-O3` flag that does not appear with optimizations disabled as of commit 693939c. The `</layer>` and `</staff>` tags appear before the last of the layer elements are written. These elements appear after the `</staff>` and are invalid.

This is the script run in node that produced the attached MEIs.
```javascript
var fs = require('fs');
eval(fs.readFileSync('verovio-toolkit.js').toString());
var mei = fs.readFileSync('mei4.mei').toString();

var options = {
    breaks: "none",
    pageHeight: 400,
    pageWidth: 600,
    noFooter: 1,
    noHeader: 1
};

var vrvToolkit = new verovio.toolkit();
vrvToolkit.setOptions(options);
vrvToolkit.loadData(mei);
console.log(vrvToolkit.getMEI(0, true));
```
[no-optimizations.txt](https://github.com/rism-ch/verovio/files/2076805/no-optimizations.txt)
[optimization-level-3.txt](https://github.com/rism-ch/verovio/files/2076806/optimization-level-3.txt)

**Comment by lpugin** [comment]
Can you post the MEI file too?

**Comment by JRegimbal** [comment]
Woops sorry!
[mei4.txt](https://github.com/rism-ch/verovio/files/2076870/mei4.txt)

**Comment by lpugin** [comment]
What version of the emdk did you use to build it? The latest version just published with 1.38.5 and with built optimization -O3 does not produce the error. At least with the current version of the MEI-viewer this is the MEI I obtain from the console with `vrvToolkit.getMEI(0, true)`:
```xml
<?xml version=\"1.0\" encoding=\"UTF-8\"?>
<?xml-model href=\"http://music-encoding.org/schema/4.0.0/mei-all.rng\" type=\"application/xml\" schematypens=\"http://relaxng.org/ns/structure/1.0\"?>
<?xml-model href=\"http://music-encoding.org/schema/4.0.0/mei-all.rng\" type=\"application/xml\" schematypens=\"http://purl.oclc.org/dsdl/schematron\"?>
<mei xmlns=\"http://www.music-encoding.org/ns/mei\" meiversion=\"4.0.0\">
    <meiHead>
        <fileDesc>
            <titleStmt>
                <title>Mensural alignment example</title>
            </titleStmt>
            <pubStmt />
            <seriesStmt>
                <title>Verovio feature examples</title>
            </seriesStmt>
        </fileDesc>
    </meiHead>
    <music>
        <body>
            <mdiv xml:id=\"mdiv-0000001753866066\">
                <score xml:id=\"score-0000001483558352\">
                    <scoreDef xml:id=\"scoredef-0000001145401580\">
                        <staffGrp xml:id=\"staffgrp-0000000155013534\" symbol=\"bracket\">
                            <staffDef xml:id=\"staffdef-0000001489623769\" clef.shape=\"C\" clef.line=\"2\" proport.num=\"1\" proport.numbase=\"3\" prolatio=\"3\" tempus=\"3\" n=\"1\" notationtype=\"neume\" lines=\"4\" />
                        </staffGrp>
                    </scoreDef>
                    <section xml:id=\"section-0000000848954190\">
                        <staff xml:id=\"staff-0000000267431697\" n=\"1\">
                            <layer xml:id=\"layer-0000001770659256\" n=\"1\">
                                <syllable xml:id=\"syllable-0000000515661715\">
                                    <syl xml:id=\"syl-0000000130627346\" n=\"initial\">
                                        <rend xml:id=\"rend-0000001140514362\" color=\"red\">O</rend>
                                    </syl>
                                    <neume xml:id=\"neume-0000000730978663\">
                                        <nc xml:id=\"nc-0000001885073740\" oct=\"3\" pname=\"e\" />
                                        <nc xml:id=\"nc-0000000587598460\" oct=\"3\" pname=\"d\" intm=\"d\" />
                                        <nc xml:id=\"nc-0000001248763792\" oct=\"3\" pname=\"e\" intm=\"u\" />
                                    </neume>
                                </syllable>
                                <syllable xml:id=\"syllable-0000000281126761\">
                                    <syl xml:id=\"syl-0000000818751889\">splen_</syl>
                                    <neume xml:id=\"neume-0000000198377428\">
                                        <nc xml:id=\"nc-0000000210333388\" oct=\"3\" pname=\"g\" />
                                        <nc xml:id=\"nc-0000000886712475\" oct=\"3\" pname=\"e\" intm=\"d\" />
                                    </neume>
                                    <neume xml:id=\"neume-0000001766765580\">
                                        <nc xml:id=\"nc-0000001693563924\" oct=\"3\" pname=\"d\" />
                                        <nc xml:id=\"nc-0000001639391192\" oct=\"3\" pname=\"e\" intm=\"u\" />
                                    </neume>
                                </syllable>
                                <syllable xml:id=\"syllable-0000001763309662\">
                                    <syl xml:id=\"syl-0000001782345441\">di_</syl>
                                    <neume xml:id=\"neume-0000001750080181\">
                                        <nc xml:id=\"nc-0000001547146234\" oct=\"3\" pname=\"f\" />
                                        <nc xml:id=\"nc-0000001269859691\" oct=\"3\" pname=\"d\" intm=\"d\" />
                                        <nc xml:id=\"nc-0000000285949630\" oct=\"3\" pname=\"c\" intm=\"d\" />
                                    </neume>
                                </syllable>
                                <syllable xml:id=\"syllable-0000002041225637\">
                                    <syl xml:id=\"syl-0000000733475490\">dis_</syl>
                                    <neume xml:id=\"neume-0000000722424460\">
                                        <nc xml:id=\"nc-0000001077595092\" oct=\"3\" pname=\"e\" />
                                    </neume>
                                </syllable>
                                <syllable xml:id=\"syllable-0000001983549378\">
                                    <syl xml:id=\"syl-0000002130921614\">si_</syl>
                                    <neume xml:id=\"neume-0000001570259142\">
                                        <nc xml:id=\"nc-0000000302487537\" oct=\"2\" pname=\"a\" />
                                        <nc xml:id=\"nc-0000002140550836\" oct=\"2\" pname=\"b\" intm=\"u\" />
                                        <nc xml:id=\"nc-0000000309171828\" oct=\"3\" pname=\"c\" intm=\"u\" />
                                    </neume>
                                </syllable>
                                <syllable xml:id=\"syllable-0000002060540759\">
                                    <syl xml:id=\"syl-0000001629032540\">ma</syl>
                                    <neume xml:id=\"neume-0000001353033927\">
                                        <nc xml:id=\"nc-0000000395947788\" oct=\"2\" pname=\"b\" />
                                        <nc xml:id=\"nc-0000000831620732\" oct=\"2\" pname=\"a\" intm=\"d\" />
                                    </neume>
                                </syllable>
                                <custos oct=\"3\" pname=\"e\" xml:id=\"custos-0000002043482625\" />
                            </layer>
                        </staff>
                    </section>
                </score>
            </mdiv>
        </body>
    </music>
</mei>
```

**Comment by JRegimbal** [comment]
I'm also using emsdk 1.38.5.

**Comment by JRegimbal** [comment]
It works fine when `vrvToolkit.setOptions(options)` is omitted.

**Comment by JRegimbal** [comment]
Okay we looked at it and found our problem disappears by disabling javascript-specific optimizations. That is replacing `-O3` with `--llvm-opts 3 --js-opts 0` so we keep the LLVM optimizations which are totally fine.

**Comment by JRegimbal** [comment]
Hmm I've been using node 8.2.1. I know that @napulen has been able to reproduce it but I'll see if I can get anyone else to do it. I'm still digging in to seeing exactly what's causing the issue for me.

**Comment by lpugin** [comment]
Bug fix 63fe0c8b1d94d013cc6a5bc88afcbe5860fbd412 should fix this.

**Comment by JRegimbal** [comment]
63fe0c8 works for me.
